components/feature/push/src/main/java/mozilla/components/feature/push/Connection.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package mozilla.components.feature.push
+
+import androidx.annotation.VisibleForTesting
+import mozilla.appservices.push.BridgeType
+import mozilla.appservices.push.PushAPI
+import mozilla.appservices.push.PushManager
+import mozilla.appservices.push.SubscriptionResponse
+import java.io.Closeable
+
+interface PushConnection : Closeable {
+    suspend fun subscribe(channelId: String, scope: String = ""): SubscriptionResponse
+    suspend fun unsubscribe(channelId: String): Boolean
+    suspend fun unsubscribeAll(): Boolean
+    suspend fun updateToken(token: String): Boolean
+    suspend fun verifyConnection(): Map<String, String>
+    fun decrypt(
+        channelId: String,
+        body: String,
+        encoding: String = "",
+        salt: String = "",
+        cryptoKey: String = ""
+    ): ByteArray
+    fun isInitialized(): Boolean
+}
+
+internal class RustPushConnection(
+    private val databasePath: String,
+    private val senderId: String,
+    private val serverHost: String,
+    private val socketProtocol: Protocol,
+    private val serviceType: ServiceType
+) : PushConnection {
+
+    @VisibleForTesting
+    internal var api: PushAPI? = null
+
+    override suspend fun subscribe(channelId: String, scope: String): SubscriptionResponse = synchronized(this) {
+        val pushApi = api
```

Comment:
This is another case of multiple folks to reviewing this change: it was recommended in order to remove the `!!` at the caller since the API is nullable (see above comments for more details on this).

The val ensures the same binding is used when calling the method's below.